Develop stream 2024-07-01

Naraenda opened the pull request:
This PR brings various updates, intended for ROCm 6.3.

It contains the following merge commits:

* 0350e0f5 perf: improve performance on rocm 6.2.
* ade3f6b2 fix: performance regression introduced with hip graph support.
* c90e74c1 docs: add documentation for hip graph support
* fa13dd77 feat: hip graph support for poisson distribution
* 4224428f feat: host implementation for mt19937
* 615fc5a7 perf: use alias method instead of binary search in cdf
* 8f7e2af7 docs: reference `GPU_TARGETS` instead of `AMDGPU_TARGET`
* 68072698 fix: fix host generators not always supporting large sizes
* 50dab4f9 refactor: remove deprecated internal headers
* 80c16d40 test: improve accuracy of poisson histogram test
* 1a44b48c docs: fix styling

This PR does not contain fixes for the recent performance regressions, we'll either add them here or create a new PR for those.

parbenc committed:
refactor mt19937 to support host version as well

parbenc committed:
update test

parbenc committed:
move jump_ahead_thread_count back to template param

parbenc committed:
implement memcpy in host and device systems

parbenc committed:
fix compilation issues and segfault

parbenc committed:
fix jump_ahead on host

parbenc committed:
create host implementation of some functions
refactor generate_long_mt19937 to work on host as well

parbenc committed:
fix remaining inconsistencies in host mt19937 generator

parbenc committed:
fix format end compile errors

Commit:
chore(gitignore): ignore python venvs

Commit:
docs(api reference): rocm-docs-core headers and stylesheets in doxyfile

rocm-docs-core distributes headers and stylesheets for doxygen for
embedding its HTML output into sphinx. These mostly fix dark-theme
and other minor visual issues when doxygen output is used this way.

Commit:
Merge branch 'fix_doxygen_style' into 'develop_stream'

docs(api reference): rocm-docs-core headers and stylesheets in doxyfile

See merge request amd/libraries/rocRAND!326

parbenc committed:
fix missing gen_next_n calls
fix clang-format issues

parbenc committed:
fix format issues and missing __host__s

nolmoonen committed:
improve accuracy of poisson histogram test

parbenc committed:
fix format and copyright dates

parbenc committed:
fix messed up host/device allocations

nolmoonen committed:
Merge branch '240-poisson-distribution-test-failure' into 'develop_stream'

Improve accuracy of Poisson histogram test

Closes #240

See merge request amd/libraries/rocRAND!327

parbenc committed:
fix merge conflicts
fix format

parbenc committed:
fix format issues and compile error

mfep committed:
Using .lint:clang-format

mfep committed:
feat(test): Added CMake option RUN_EXTRA_TESTS

mfep committed:
fix(generator): Fixed the usage of min in host generators

NB4444 committed:
Removed deprecated internal headers, src/rng/distribution/distributions.hpp and src/rng/device_engines.hpp

NB4444 committed:
Merge branch '341-remove-deprecated-internal-headers' into 'develop_stream'

Resolve "Remove deprecated internal headers"

Closes #341

See merge request amd/libraries/rocRAND!330

parbenc committed:
fix format issues

mfep committed:
feat(test): Added large size tests for host generators

mfep committed:
Merge branch '346-host-large-sizes' into 'develop_stream'

Resolve "Some host generators might not support large sizes due to min / max"

See merge request amd/libraries/rocRAND!329

parbenc committed:
fix format issue

parbenc committed:
disable most mt19937 host tests for normal run (enabled for slow test run)

Commit:
docs(dyn_ordering): Use GPU_TARGETS instead of AMDGPU_TARGETS

AMDGPU_TARGETS doesn't pick up updates correctly (needs cache clean) whereas GPU_TARGETS does. Every other doc and CI too refers to GPU_TARGETS.

parbenc committed:
fix review comments

Commit:
Merge branch 'docs-gpu-targets' into 'develop_stream'

docs(dyn_ordering): Use GPU_TARGETS instead of AMDGPU_TARGETS

See merge request amd/libraries/rocRAND!332

ex-rzr committed:
Use alias method in rocrand_discrete for MTGP32, LFSR113 and ThreeFry

discrete_alias is faster than discrete_cdf. Though discrete_cdf can be used with PRNGs,
it is supposed to be used with QRNGs (Sobol generators) as it maintains quasi-randomness.

ex-rzr committed:
Merge branch '285-discrete_cdf-is-incorrectly-used-by-some-prngs' into 'develop_stream'

Use discrete_alias instead of discrete_cdf with all PRNGs

Closes #285

See merge request amd/libraries/rocRAND!333

mfep committed:
Removed 'apt-get install flang'

parbenc committed:
remove synchronization from host_system::launch function
add synchronization to host_system alloc, free and memcpy

mfep committed:
use ROCRAND_HIP_FATAL_ASSERT for hipDeviceSynchronize call

parbenc committed:
Merge branch '309_mt19937_host' into 'develop_stream'

create host implementation for mt19937 generator

See merge request amd/libraries/rocRAND!325

mfep committed:
Removed accidentally included #include "hip/amd_detail/host_defines.h"

mfep committed:
Implement asynchronous initialization of poisson distribution

mfep committed:
generate_poisson test with many lambdas

mfep committed:
Testing poisson with hipGraphs

mfep committed:
Test [blocking] host_generator with non-blocking stream

mfep committed:
Fixing poisson distribution selection in benchmark_tuning

mfep committed:
Updated changelog

mfep committed:
Merge branch 'poisson-graph-2' into 'develop_stream'

Support for poisson distribution in graphs

Closes #361

See merge request amd/libraries/rocRAND!334

mfep committed:
fix(docs): Added links to unaccessible doc pages

mfep committed:
fix(docs): Removed duplicated CUDA Compatibility section from Programmer's guide

mfep committed:
Added hipGraphs doc and sample

ex-rzr committed:
Fix performance regression of Poisson distribution

Recent changes required for HIP graph support added a new path with
approximation of Poisson with normal distribution when lambda is large.
However, the decision whether to use the alias/CDF methods or the
approximation is made in the kernel for every generated value even
though lambda is the same.
This change moves it to host side: depending on lambda the kernel is
launched with one of two distributions (poisson_distribution or
poisson_distribution_huge).

ex-rzr committed:
Merge branch '360-doc-hip-graph' into 'develop_stream'

Resolve "Document HIP Graph support"

Closes #360

See merge request amd/libraries/rocRAND!335

ex-rzr committed:
Merge branch '366-fix-performance-regression-of-poisson' into 'develop_stream'

Fix performance regression of Poisson distribution introduced by HIP graph support

Closes #366

See merge request amd/libraries/rocRAND!336

ex-rzr committed:
clang-format: Break after attributes

ex-rzr committed:
Add missing __forceinline__ to improve performance on ROCm 6.2

hipcc from ROCm 6.2 does not add `-mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false`
by default.

ex-rzr committed:
Remove meaningless code in xorwow introduced during rebase/merge

ex-rzr committed:
Merge branch '363-improve-performance-on-rocm-6-2' into 'develop_stream'

Improve performance on ROCm 6.2

See merge request amd/libraries/rocRAND!337

Naraenda commented:
@samjwu The clang format CI step seems to not handle external PRs properly, I think this might also be happening in other repositories.

Naraenda committed:
style: update formatting

Naraenda committed:
ci(.gitlab-ci.yml): replace 'ROCM_PATH' variable with 'env:HIP_PATH' as the former is unintuitive

Naraenda committed:
ci(.gitlab-ci.yml): do not force download deps on windows

Naraenda committed:
ci(.gitlab-ci.yml): pass amdclang filepath properly to windows package build test

Naraenda committed:
Merge branch 'ci-remove-rocm-path-variable' into 'develop_stream'

ci(.gitlab-ci.yml): replace 'ROCM_PATH' variable with 'env:HIP_PATH' as the former is unintuitive

See merge request amd/libraries/rocRAND!339

Naraenda commented:
@stanleytsang-amd Let me know if I should squash the commits into their respective merge commits.

stanleytsang-amd commented:
> @samjwu The clang format CI step seems to not handle external PRs properly, I think this might also be happening in other repositories.
@samjwu Can you take a look at the clang-format CI?

Run git fetch origin develop
From https://github.com/ROCm/rocRAND
 * branch            develop    -> FETCH_HEAD
 * [new branch]      develop    -> origin/develop
fatal: couldn't find remote ref develop_stream_20240701
Error: Process completed with exit code [12](https://github.com/ROCm/rocRAND/actions/runs/9746751218/job/26897847285?pr=512#step:6:13)8.

Naraenda commented:
Rebased to get the clang-format ci fixes.

NB4444 committed:
Added checks for nullptr data with tests

ex-rzr committed:
Remove unused FindTestU01.cmake

ex-rzr committed:
Merge branch 'remove-unused-findtestu01-cmake' into 'develop_stream'

Remove unused FindTestU01.cmake

See merge request amd/libraries/rocRAND!342

ex-rzr committed:
Fix bit rotation for threefry2x64 and threefry4x64

operator >> has higher precedence than operator &.
This bug causes very low quality in crush tests.

NB4444 committed:
Merge branch '369-add-check-for-nullptr-data-when-calling-host-generate' into 'develop_stream'

Resolve "Add check for nullptr data when calling host generate."

Closes #369

See merge request amd/libraries/rocRAND!341

ex-rzr committed:
Merge branch '371-fix-threefry2x64-and-threefry4x64' into 'develop_stream'

Fix threefry2x64 and threefry4x64

Closes #371

See merge request amd/libraries/rocRAND!343

Naraenda commented:
Rebased and added threefry fixes by @ex-rzr

umfranzw reviewed (approved):
I've finished working my way through this one, and I think it looks good. All our CI checks are also passing now.
@stanleytsang-amd if you have no objections, I'm ok with this one going in.

stanleytsang-amd commented on CMakeLists.txt:
```diff
@@ -33,6 +33,7 @@ cmake_dependent_option(BUILD_BENCHMARK_TUNING
 option(BUILD_ADDRESS_SANITIZER "Build with address sanitizer enabled" OFF)
 option(CODE_COVERAGE "Build with code coverage flags (clang only)" OFF)
 option(DEPENDENCIES_FORCE_DOWNLOAD "Don't search the system for dependencies, always download them" OFF)
+cmake_dependent_option(RUN_SLOW_TESTS "Run extra tests with CTest. These cover niche functionality and take long time" OFF "BUILD_TEST" OFF)
```
Please update the packaging version in CMakeLists.txt to match the changelog.

Naraenda replied:
Done in ec409a3

stanleytsang-amd replied:
thanks

stanleytsang-amd reviewed (changes requested):
LGTM once the packaging version is updated.

Naraenda committed:
chore: bump version
